pallets/capacity/src/lib.rs

```diff
@@ -126,29 +142,138 @@ pub mod pallet {
 
 	#[pallet::event]
 	#[pallet::generate_deposit(pub (super) fn deposit_event)]
-	pub enum Event<T: Config> {}
+	pub enum Event<T: Config> {
+		/// Tokens have been staked to the Frequency network.
+		Staked {
+			/// The token account that staked tokens to the network.
+			account: T::AccountId,
+			/// The MSA that a token account targeted to receive Capacity based on this staking amount.
+			target: MessageSourceId,
+			/// An amount that was staked.
+			amount: BalanceOf<T>,
+		},
+	}
 
 	#[pallet::error]
-	pub enum Error<T> {}
+	pub enum Error<T> {
+		/// Staker attempted to stake to an invalid staking target.
+		InvalidTarget,
+		/// Staker has insufficient balance to cover the amount wanting to stake.
+		InsufficientBalance,
+		/// Staker is attempting to stake an amount below the minimum amount.
+		InsufficientStakingAmount,
+		/// Staker is attempting to stake a zero amount.
+		ZeroAmountNotAllowed,
+	}
 
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
 		/// Stakes some amount of tokens to the network and generates Capacity.
 		///
 		/// ### Errors
 		///
-		/// - Returns Error::InsufficientBalance if the sender does not have free balance amount needed to stake.
+		/// - Returns Error::ZeroAmountNotAllowed if the staker is attempting to stake a zero amount.
 		/// - Returns Error::InvalidTarget if attempting to stake to an invalid target.
 		/// - Returns Error::InsufficientStakingAmount if attempting to stake an amount below the minimum amount.
 		#[pallet::weight(T::WeightInfo::stake())]
 		pub fn stake(
-			_origin: OriginFor<T>,
-			_target: MessageSourceId,
-			_amount: BalanceOf<T>,
+			origin: OriginFor<T>,
+			target: MessageSourceId,
+			amount: BalanceOf<T>,
 		) -> DispatchResult {
+			let staker = ensure_signed(origin)?;
+
+			let (mut staking_account, actual_amount) =
+				Self::ensure_can_stake(&staker, target, amount)?;
+
+			Self::increase_stake_and_issue_capacity(
+				&staker,
+				&mut staking_account,
+				target,
+				actual_amount,
+			);
+
+			Self::deposit_event(Event::Staked { account: staker, amount: actual_amount, target });
+
 			Ok(())
 		}
 	}
 }
 
-impl<T: Config> Pallet<T> {}
+impl<T: Config> Pallet<T> {
+	/// Checks to see if staker has sufficient free-balance to stake the minimum required staking amount.
+	///
+	/// # Errors
+	/// * [`Error::ZeroAmountNotAllowed`]
+	/// * [`Error::InvalidTarget`]
+	///
+	fn ensure_can_stake(
+		staker: &T::AccountId,
+		target: MessageSourceId,
+		amount: BalanceOf<T>,
+	) -> Result<(StakingAccountDetails<T>, BalanceOf<T>), DispatchError> {
+		ensure!(amount > Zero::zero(), Error::<T>::ZeroAmountNotAllowed);
+		ensure!(T::TargetValidator::validate(target), Error::<T>::InvalidTarget);
+
+		let staking_account = Self::get_staking_account_for(&staker).unwrap_or_default();
+		let stakable_amount = staking_account.get_stakable_amount_for(&staker, amount);
```

Comment:
Should the `ZeroAmountNotAllowed` check be here instead?
```suggestion
		let stakable_amount = staking_account.get_stakable_amount_for(&staker, amount);
		ensure!(stakable_amount > Zero::zero(), Error::<T>::ZeroAmountNotAllowed);
```

Reply:
Or perhaps we don't need it with the `InsufficientStakingAmount`?

Reply:
They would have needed some free balance to submit this transaction and get this far. So this means that the Staker has some amount to stake, whether it's a tiny amount left on their account.

If the Staker attempts to stake an amount smaller than the staking amount on the first staking attempt, that will create an error.

If the Staker, on subsequent attempts (re-staking), tries to stake an additional amount, no matter how small the amount is, it will be allowed.